Table: Add support for wrapping text

bradley-eb opened the pull request:
## Description
The table should allow for the text to wrap, instead of being truncated. Since the current table is setup to truncate by default, we should leave that. My suggestion is to add a wrap-text property that defaults to false. When enabled, it will remove the css that currently causes the body text to be truncated with ellipsis. I do not expect this setting to affect the header or the footer.

References #1998

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

<!--Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

dnarey commented:
Feedback from UX:
Here are the expectations
1. The checkboxes should be top aligned
2. The Text/anything in all the cells should be top aligned
3. Hyperlinks also have to be truncated.

apaddock commented:
@bradley-eb fix the following issues (per ux) 

- [ ]  The checkboxes should be top aligned

- [ ] The Text/anything in all the cells should be top aligned

- [ ] Hyperlinks also have to be truncated.

JTrimbleR commented:
Validated test scenarios after feedback, working as expected.

Pre-conditions: wrapText control set to true, columnResize control set to true

Scenario 1: Resize a column which text’s length exceeds the min width to min width, text should wrap (no truncation) and all other field in the row also resize

Scenario 2: Resize a column which text’s length exceeds the max width to max width, text should wrap (no truncation) and all other field in the row also resize

Scenario 3: Validate that columns that have text with special characters are able to wrap (&, %, “, ')

Scenario 4: Validate that a column whose text is wrapping does not affect the size of other columns

Scenario 5: Checkboxes

Scenario 6: Badges

Scenario 7: Column Resize

Scenario 8: Hyperlinks

Scenario 9: Borderless table

Scenario 10: Value Formatter

Scenario 11: Pagination

Scenario 12: Expandable rows

Scenario 13: Densities

cjwinsor commented on stencil-workspace/src/components/modus-table/parts/cell/modus-table-cell.tsx:
```diff
@@ -23,6 +23,7 @@ export const ModusTableCell: FunctionalComponent<ModusTableCellProps> = ({ cell,
       class={`
       ${hasRowsExpandable || frozenColumns.includes(cell.column.id) ? 'sticky-left' : ''}
       ${cell.column.getIsResizing() ? 'active-resize' : ''}
+      ${wrapText ? 'align-top' : ''}
```
We shouldn't have to add a css class to every `td`.  Ideally no extra class is needed to do the padding with or without text-wrap.  Given the density class and wrapped text classes though, you should be able to manage without the extra class.

Below is something just to give an idea of how it can be done.
```scss
table {
  &.density-relaxed {
    &.truncate-text tbody tr td {
      //...
    }
    
    &.wrap-text tbody tr td {
     //...
    }
  }
}
```

bradley-eb replied:
The classname and property `align-top` has been removed.

cjwinsor reviewed (changes requested):
Make sure this pr only has your changes (i see a change to the readme of modus-icons).  I would like to avoid adding classes to every cell of tbody.

bradley-eb commented:
> Make sure this pr only has your changes (i see a change to the readme of modus-icons). I would like to avoid adding classes to every cell of tbody.

@cjwinsor Thanks for your review. I have removed unnecessary changes, fixed all lint issues as well, and got rid of `align-top`.

JTrimbleR commented:
Validated scenarios against latest changes, working as expected.

Pre-conditions: wrapText control set to true, columnResize control set to true

Scenario 1: Resize a column which text’s length exceeds the min width to min width, text should wrap (no truncation) and all other field in the row also resize

Scenario 2: Resize a column which text’s length exceeds the max width to max width, text should wrap (no truncation) and all other field in the row also resize

Scenario 3: Validate that columns that have text with special characters are able to wrap (&, %, “, ')

Scenario 4: Validate that a column whose text is wrapping does not affect the size of other columns

Scenario 5: Checkboxes

Scenario 6: Badges

Scenario 7: Column Resize

Scenario 8: Hyperlinks

Scenario 9: Borderless table

Scenario 10: Value Formatter

Scenario 11: Pagination

Scenario 12: Expandable rows

Scenario 13: Densities

cjwinsor commented on stencil-workspace/src/components/modus-table/modus-table.scss:
```diff
@@ -197,16 +197,31 @@ table {
+      .truncate-text {
         overflow: hidden;
         text-overflow: ellipsis;
         text-wrap: nowrap;
```
Change `text-wrap: nowrap` to `white-space: nowrap`, this will be important for the test update suggestion I am going to make as `text-wrap` isn't a fully supported property so my test suggestion won't work with it.  Also the rest of MWC uses `white-space: nowrap`.

bradley-eb replied:
Updated.

cjwinsor commented on stencil-workspace/src/components/modus-table/modus-table.e2e.ts:
```diff
@@ -935,4 +935,28 @@ describe('modus-table', () => {
+    const tableData = await page.findAll('modus-table >>> td');
+    const firstItem = tableData[0].textContent;
+
+    expect(firstItem).toBe('This is an example of long text');
```
This will always assert true as the content isn't affected by the truncation. This is only visual.

My suggestion is to update the test so it validates that `wrap-text` is applied and that the styles are correct.  We also need another test that validates when `wrapText` is false.

Write a selector that finds the span, so you can check `wrapText = false`
```js
    const style = await spanData.getComputedStyle();
    expect(style['overflow']).toBe('hidden');
    expect(style['text-overflow']).toBe('ellipsis');
    expect(style['white-space']).toBe('nowrap');
```

and for `wrapText = true`

```js
   const style = await spanData.getComputedStyle();
    expect(style['overflow-wrap']).toBe('break-word');
    expect(style['word-break']).toBe('break-word');
```
